# Exercise I

# Exercise II
Using single unit propeties from the 2017 oberstabations:

### Data Aquisition: 

1. Acquire bedroomcnt, bathroomcnt, calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, and fips from the zillow database for all 'Single Family Residential' properties.

In [1]:
# imports 
import pandas as pd
import numpy as np
import env

In [2]:
# need to make a funciton to aquire the url connection
def get_db_url(db, user = env.user, host = env.host, password = env.password):
    '''
    This function will take in a databse and set enviromental credentials and
    return a datbase connection link. 
    '''
    return f'mysql+pymysql://{user}:{password}@{host}/{db}'

In [3]:
url = get_db_url('zillow')
#url

In [4]:
# need to make query that pulls all the information that is requested
sql_query = '''
select 
    propertylandusetypeid,
    propertylandusedesc,
    bedroomcnt,
    bathroomcnt,
    calculatedfinishedsquarefeet,
    taxvaluedollarcnt,
    yearbuilt,
    taxamount,
    fips
from properties_2017
Left join propertylandusetype as pt
    Using (propertylandusetypeid)
where pt.propertylandusedesc like 'Single%%'
    '''

In [5]:
# a different sequel query that returns the same results but without the join:
sql_query2 = '''
Select 
	propertylandusetypeid,
	bedroomcnt,
    bathroomcnt,
    calculatedfinishedsquarefeet,
    taxvaluedollarcnt,
    yearbuilt,
    taxamount,
    fips
from properties_2017 
where propertylandusetypeid = 261;
'''

In [6]:
# make sure the query works
df = pd.read_sql(sql_query, get_db_url('zillow'))
df.head()

propertylandusetypeid        propertylandusedesc  bedroomcnt  bathroomcnt  \
0                  261.0  Single Family Residential         0.0          0.0   
1                  261.0  Single Family Residential         0.0          0.0   
2                  261.0  Single Family Residential         0.0          0.0   
3                  261.0  Single Family Residential         0.0          0.0   
4                  261.0  Single Family Residential         4.0          2.0   

   calculatedfinishedsquarefeet  taxvaluedollarcnt  yearbuilt  taxamount  \
0                           NaN            27516.0        NaN        NaN   
1                           NaN               10.0        NaN        NaN   
2                           NaN               10.0        NaN        NaN   
3                           NaN             2108.0        NaN     174.21   
4                        3633.0           296425.0     2005.0    6941.39   

     fips  
0  6037.0  
1  6037.0  
2  6037.0  
3  6037.0  
4  6037.0

In [7]:
# make a function to get the data using connection url
def new_zillow_data():
    '''
    This function will use a set sql query and return a zillow data frame based on the 
    exercise II in the ds.codeup.com regression: acquistion and prep request
    '''
    
    sql_query = '''
                select
                    id,
                    parcelid,
                    propertylandusetypeid,
                    propertylandusedesc,
                    bedroomcnt,
                    bathroomcnt,
                    calculatedfinishedsquarefeet,
                    taxvaluedollarcnt,
                    yearbuilt,
                    taxamount,
                    fips
                from properties_2017
                Left join propertylandusetype as pt
                    Using (propertylandusetypeid)
                where pt.propertylandusedesc like 'Single%%'
                '''
    df = pd.read_sql(sql_query, get_db_url('zillow'))
    
    return df

In [8]:
#check that it works
df = new_zillow_data()
df.head()

id  parcelid  propertylandusetypeid        propertylandusedesc  bedroomcnt  \
0   1  10759547                  261.0  Single Family Residential         0.0   
1  15  11193347                  261.0  Single Family Residential         0.0   
2  16  11215747                  261.0  Single Family Residential         0.0   
3  17  11229347                  261.0  Single Family Residential         0.0   
4  20  11324547                  261.0  Single Family Residential         4.0   

   bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  yearbuilt  \
0          0.0                           NaN            27516.0        NaN   
1          0.0                           NaN               10.0        NaN   
2          0.0                           NaN               10.0        NaN   
3          0.0                           NaN             2108.0        NaN   
4          2.0                        3633.0           296425.0     2005.0   

   taxamount    fips  
0        NaN  6037.0  
1        NaN  6037.0  
2        NaN  6037.0  
3     174.21  6037.0  
4    6941.39  6037.0

In [9]:
import os

In [10]:
# MAKE IT INTO A CSV:
def get_zillow_data():
    '''
    This function reads in data from a codeup database, writes the data to a csv file if a 
    local file does not exist, and returns a df
    '''
    
    if os.path.isfile('zillow.csv'):
        
        #if csv file exists, read in data from csv file 
        df = pd.read_csv('zillow.csv', index_col = 0)
        
    else:
        
        # read fresh data from db into a dataframe
        df = new_zillow_data()
        
        #write dataframe to a csv file
        df.to_csv('zillow.csv')
    
    return df

In [11]:
#check that it works
df = get_zillow_data()
df.head()

id  parcelid  propertylandusetypeid        propertylandusedesc  bedroomcnt  \
0   1  10759547                  261.0  Single Family Residential         0.0   
1  15  11193347                  261.0  Single Family Residential         0.0   
2  16  11215747                  261.0  Single Family Residential         0.0   
3  17  11229347                  261.0  Single Family Residential         0.0   
4  20  11324547                  261.0  Single Family Residential         4.0   

   bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  yearbuilt  \
0          0.0                           NaN            27516.0        NaN   
1          0.0                           NaN               10.0        NaN   
2          0.0                           NaN               10.0        NaN   
3          0.0                           NaN             2108.0        NaN   
4          2.0                        3633.0           296425.0     2005.0   

   taxamount    fips  
0        NaN  6037.0  
1        NaN  6037.0  
2        NaN  6037.0  
3     174.21  6037.0  
4    6941.39  6037.0

### Data Cleaning: (Prepare)

2. Using your acquired Zillow data, walk through the summarization and cleaning steps in your wrangle.ipynb file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.

In [12]:
df.shape

(2152863, 11)

In [13]:
df.info(show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2152863 entries, 0 to 2152862
Data columns (total 11 columns):
 #   Column                        Non-Null Count    Dtype  
---  ------                        --------------    -----  
 0   id                            2152863 non-null  int64  
 1   parcelid                      2152863 non-null  int64  
 2   propertylandusetypeid         2152863 non-null  float64
 3   propertylandusedesc           2152863 non-null  object 
 4   bedroomcnt                    2152852 non-null  float64
 5   bathroomcnt                   2152852 non-null  float64
 6   calculatedfinishedsquarefeet  2144379 non-null  float64
 7   taxvaluedollarcnt             2152370 non-null  float64
 8   yearbuilt                     2143526 non-null  float64
 9   taxamount                     2148421 non-null  float64
 10  fips                          2152863 non-null  float64
dtypes: float64(8), int64(2), object(1)
memory usage: 180.7+ MB


In [14]:
df.describe()

id      parcelid  propertylandusetypeid    bedroomcnt  \
count  2.152863e+06  2.152863e+06              2152863.0  2.152852e+06   
mean   1.490834e+06  1.309657e+07                  261.0  3.287196e+00   
std    8.608785e+05  5.171557e+06                    0.0  9.547544e-01   
min    1.000000e+00  1.071172e+07                  261.0  0.000000e+00   
25%    7.451170e+05  1.165907e+07                  261.0  3.000000e+00   
50%    1.490613e+06  1.259852e+07                  261.0  3.000000e+00   
75%    2.235860e+06  1.405609e+07                  261.0  4.000000e+00   
max    2.982282e+06  1.696019e+08                  261.0  2.500000e+01   

        bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
count  2.152852e+06                  2.144379e+06       2.152370e+06   
mean   2.230688e+00                  1.862855e+03       4.618962e+05   
std    9.992796e-01                  1.222125e+03       6.996760e+05   
min    0.000000e+00                  1.000000e+00       1.000000e+00   
25%    2.000000e+00                  1.257000e+03       1.881702e+05   
50%    2.000000e+00                  1.623000e+03       3.276710e+05   
75%    3.000000e+00                  2.208000e+03       5.345270e+05   
max    3.200000e+01                  9.525760e+05       9.842891e+07   

          yearbuilt     taxamount          fips  
count  2.143526e+06  2.148421e+06  2.152863e+06  
mean   1.960950e+03  5.634866e+03  6.048377e+03  
std    2.216220e+01  8.178910e+03  2.043329e+01  
min    1.801000e+03  1.850000e+00  6.037000e+03  
25%    1.949000e+03  2.534980e+03  6.037000e+03  
50%    1.958000e+03  4.108950e+03  6.037000e+03  
75%    1.976000e+03  6.414320e+03  6.059000e+03  
max    2.016000e+03  1.337756e+06  6.111000e+03

Initial observations: 
- There are 2,152,863 columns and 9 rows
- The row names need to be fixed
- All the data types will be floats once I drop propetylandusedesc
- There are alot of nulls in the data. 🥹

In [15]:
# look at all the value counts
df.isnull().sum()

id                                 0
parcelid                           0
propertylandusetypeid              0
propertylandusedesc                0
bedroomcnt                        11
bathroomcnt                       11
calculatedfinishedsquarefeet    8484
taxvaluedollarcnt                493
yearbuilt                       9337
taxamount                       4442
fips                               0
dtype: int64

In [16]:
df[df.yearbuilt.isnull()]

id  parcelid  propertylandusetypeid        propertylandusedesc  \
0              1  10759547                  261.0  Single Family Residential   
1             15  11193347                  261.0  Single Family Residential   
2             16  11215747                  261.0  Single Family Residential   
3             17  11229347                  261.0  Single Family Residential   
5             27  11457547                  261.0  Single Family Residential   
...          ...       ...                    ...                        ...   
2152823  2982202  13889389                  261.0  Single Family Residential   
2152844  2982239  17204623                  261.0  Single Family Residential   
2152854  2982262  11477607                  261.0  Single Family Residential   
2152857  2982267  14359297                  261.0  Single Family Residential   
2152860  2982272  14356183                  261.0  Single Family Residential   

         bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
0               0.0          0.0                           NaN   
1               0.0          0.0                           NaN   
2               0.0          0.0                           NaN   
3               0.0          0.0                           NaN   
5               0.0          0.0                           NaN   
...             ...          ...                           ...   
2152823         0.0          0.0                           NaN   
2152844         0.0          0.0                           NaN   
2152854         0.0          0.0                           NaN   
2152857         0.0          0.0                           NaN   
2152860         0.0          0.0                           NaN   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  
0                  27516.0        NaN        NaN  6037.0  
1                     10.0        NaN        NaN  6037.0  
2                     10.0        NaN        NaN  6037.0  
3                   2108.0        NaN     174.21  6037.0  
5                    124.0        NaN        NaN  6037.0  
...                    ...        ...        ...     ...  
2152823          2568893.0        NaN   27309.30  6059.0  
2152844            92679.0        NaN    1090.16  6111.0  
2152854          1198476.0        NaN        NaN  6037.0  
2152857          1087111.0        NaN   19313.08  6059.0  
2152860           208057.0        NaN    5783.88  6059.0  

[9337 rows x 11 columns]